examples/all-clusters-app/linux/main-common.cpp

```diff
@@ -205,142 +214,116 @@ void HandleSwitchEvent(intptr_t arg)
+    }
+    else if (strncmp("ShortRelease", payload, strlen("ShortRelease")) == 0)
+    {
+        HandleSwitchEvent(Clusters::Switch::Events::ShortRelease::Id);
+    }
+    else if (strncmp("LongRelease", payload, strlen("LongRelease")) == 0)
+    {
+        HandleSwitchEvent(Clusters::Switch::Events::LongRelease::Id);
+    }
+    else if (strncmp("MultiPressOngoing", payload, strlen("MultiPressOngoing")) == 0)
+    {
+        HandleSwitchEvent(Clusters::Switch::Events::MultiPressOngoing::Id);
+    }
+    else if (strncmp("MultiPressComplete", payload, strlen("MultiPressComplete")) == 0)
+    {
+        HandleSwitchEvent(Clusters::Switch::Events::MultiPressComplete::Id);
+    }
+    else if (strncmp("PowerOnReboot", payload, strlen("PowerOnReboot")) == 0)
+    {
+        HandleRebootSignal(BootReasonType::kPowerOnReboot);
+    }
+    else if (strncmp("BrownOutReset", payload, strlen("BrownOutReset")) == 0)
+    {
+        HandleRebootSignal(BootReasonType::kBrownOutReset);
+    }
+    else if (strncmp("SoftwareWatchdogReset", payload, strlen("SoftwareWatchdogReset")) == 0)
+    {
+        HandleRebootSignal(BootReasonType::kSoftwareWatchdogReset);
+    }
+    else if (strncmp("HardwareWatchdogReset", payload, strlen("HardwareWatchdogReset")) == 0)
+    {
+        HandleRebootSignal(BootReasonType::kHardwareWatchdogReset);
+    }
+    else if (strncmp("SoftwareUpdateCompleted", payload, strlen("SoftwareUpdateCompleted")) == 0)
+    {
+        HandleRebootSignal(BootReasonType::kSoftwareUpdateCompleted);
+    }
+    else if (strncmp("SoftwareReset", payload, strlen("SoftwareReset")) == 0)
+    {
+        HandleRebootSignal(BootReasonType::kSoftwareReset);
+    }
+    else
+    {
+        ChipLogError(NotSpecified, "Unhandled command: Should never happens");
+    }
 }
 
-void OnGeneralFaultSignalHandler(int signum)
+void * EventCommandListenerTask(void * arg)
 {
-    ChipLogDetail(DeviceLayer, "Caught signal %d", signum);
+    int fd;
+    char readbuf[CHIP_EVENT_COMMAND_SIZE];
+    ssize_t readBytes;
 
-    uint32_t eventId;
-    switch (signum)
+    for (;;)
     {
-    case SIGUSR2:
-        eventId = Clusters::GeneralDiagnostics::Events::HardwareFaultChange::Id;
-        break;
-    case SIGHUP:
-        eventId = Clusters::GeneralDiagnostics::Events::RadioFaultChange::Id;
-        break;
-    case SIGTTIN:
-        eventId = Clusters::GeneralDiagnostics::Events::NetworkFaultChange::Id;
-        break;
-    default:
-        ChipLogError(NotSpecified, "Unhandled signal: Should never happens");
-        chipDie();
-        break;
-    }
+        fd = open(CHIP_EVENT_FIFO_PATH, O_RDONLY);
 
-    PlatformMgr().ScheduleWork(HandleGeneralFaultEvent, static_cast<intptr_t>(eventId));
-}
+        readBytes          = read(fd, readbuf, CHIP_EVENT_COMMAND_SIZE);
+        readbuf[readBytes] = '\0';
+        ChipLogProgress(Zcl, "Received payload: \"%s\" and length is %ld\n", readbuf, readBytes);
```

Comment:
The open/read/close loop could be replaced with a single open and a select loop. If this blocks right now, let's leave it as-is. I just want to make sure we don't busy-spin here.

Reply:
No, blocking occurs on a FIFO at open call. In other words, if the FIFO is opened for reading, the process will "block" until some other process opens it for writing. 

From the Open Group page for the [open()](http://pubs.opengroup.org/onlinepubs/007908799/xsh/open.html) function

O_NONBLOCK

    When opening a FIFO with O_RDONLY or O_WRONLY set: If O_NONBLOCK is
    set:

        An open() for reading only will return without delay. An open()
        for writing only will return an error if no process currently
        has the file open for reading.

    If O_NONBLOCK is clear:

        An open() for reading only will block the calling thread until a
        thread opens the file for writing. An open() for writing only
        will block the calling thread until a thread opens the file for
        reading.